torch/utils/data/dataloader.py

```diff
@@ -1191,30 +1191,36 @@ def _next_data(self):
                         self._workers_status[data.worker_id] = False
                     else:
                         self._mark_worker_as_unavailable(data.worker_id)
+                    # Given worker is exhausted, use _try_put_index to put data by looking for other available workers
                     self._try_put_index()
                     continue
 
+            worker_id = self._task_info[idx][0]
             if idx != self._rcvd_idx:
                 # store out-of-order samples
+                self._try_put_index(worker_id)
```

Comment:
why do we need to call this? Our logic only changes which woker we pick right?

Reply:
Earlier _try_put_index was called inside _process_data function. _process_data is called when data from worker is being sent outside dataloader. 


If we call _try_put_index inside _process_data function it means that we will give data to worker only when it's data processed, until then that worker will keep waiting. 

To be able to give next batch of indices to worker immediately after it fetches it data, we need to move  _try_put_index outside process data function.